adapters/beachfront/beachfront.go

```diff
@@ -0,0 +1,541 @@
+	Price float64 `json:"price"`
+	W     uint64  `json:"w"`
+	H     uint64  `json:"h"`
+	Slot  string  `json:"slot"`
+	Adm   string  `json:"adm"`
+}
+
+func (a *BeachfrontAdapter) Name() string {
+	return "beachfront"
+}
+
+func (a *BeachfrontAdapter) SkipNoCookies() bool {
+	return true
+}
+
+func (a *BeachfrontAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp))
+
+	var beachfrontRequests BeachfrontRequests
+	var reqJSON []byte
+
+	a.URI = func() string {
+		for i := range request.Imp {
+			if request.Imp[i].Video != nil {
+				// If there are any video imps, we will be running a video auction
+				// and dropping all of the banner actions.
+				return VideoEndpoint
+			}
+		}
+		return BannerEndpoint
+	}()
+
+	beachfrontRequests, err := preprocess(request, a.URI)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	if a.URI == VideoEndpoint {
+		reqJSON, err = json.Marshal(beachfrontRequests.Video)
+		a.URI = a.URI + beachfrontRequests.Video.AppId + VideoEndpointSuffix
+	} else {
+		reqJSON, err = json.Marshal(beachfrontRequests.Banner)
+	}
+
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	if a.URI == BannerEndpoint {
+		if request.User != nil {
+			headers.Add("Cookie", "UserID="+request.User.ID+"; __io_cid="+request.User.BuyerUID)
+		}
+	}
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     a.URI,
+		Body:    reqJSON,
+		Headers: headers,
+	}}, errs
+}
+
+/*
+We have received a prebid request. It needs to be converted to a beachfront request. This is complicated
+by the fact that we have different servers for video/display and they have different contracts.
+*/
+func preprocess(req *openrtb.BidRequest, uri string) (BeachfrontRequests, error) {
+	var beachfrontReqs BeachfrontRequests
+	var err error
+
+	if uri == BannerEndpoint {
+		beachfrontReqs.Banner, err = getBannerRequest(req)
+		if err != nil {
+			return beachfrontReqs, err
+		}
+	} else {
+		beachfrontReqs.Video, err = getVideoRequest(req)
+		if err != nil {
+			return beachfrontReqs, err
+		}
+	}
+
+	return beachfrontReqs, err
+}
+
+func getBannerRequest(req *openrtb.BidRequest) (BeachfrontBannerRequest, error) {
+	var beachfrontReq BeachfrontBannerRequest
+	dec := json.NewDecoder(strings.NewReader(beachfrontBannerRequestTemplate))
+	for {
+		if err := dec.Decode(&beachfrontReq); err == io.EOF {
+			break
+		} else if err != nil {
+			// possible banner error 1 - decoding error
+			return beachfrontReq, err
+		}
+	}
+
+	// step through the prebid request "imp" and inject into the beachfront request.
+	for i, imp := range req.Imp {
+		if imp.Video != nil {
+			return beachfrontReq, nil
+		} else if imp.Banner != nil {
+			if i > 0 {
+				beachfrontReq.Slots = append(beachfrontReq.Slots, BeachfrontSlot{})
```

Comment:
what happens here if there is a native `Imp` as part of the request? You expand the size of `beachfrontReq.Slots`, by 1 in order to have have the slot index available, but then reference `i` directly for the slot. If there was a native (or something else not video or banner) inbetween, `i` will have advanced by 2 since the last append, and you will get an array out of bounds error.

There are two solutions, start with setting `Slots` to be an array of size `len(req.Imp)`, or reference the slot as `beachfrontReq.Slots[len(beachfrontReq.Slots)-1]`

Also from the way you have defined your template, the first slot will have an extra 0x0 size in the array, since it won't be an empty `beachfrontSlot{}` like you are appending for all slots beyond the first.